src/common/catalog/catalog-entity.ts

```diff
@@ -0,0 +1,135 @@
+import { EventEmitter } from "events";
+import { observable } from "mobx";
+
+type ExtractEntityMetadataType<Entity> = Entity extends CatalogEntity<infer Metadata> ? Metadata : never;
+type ExtractEntityStatusType<Entity> = Entity extends CatalogEntity<any, infer Status> ? Status : never;
+type ExtractEntitySpecType<Entity> = Entity extends CatalogEntity<any, any, infer Spec> ? Spec : never;
+
+export type CatalogEntityConstructor<Entity extends CatalogEntity> = (
+  (new (data: CatalogEntityData<
+    ExtractEntityMetadataType<Entity>,
+    ExtractEntityStatusType<Entity>,
+    ExtractEntitySpecType<Entity>
+  >) => Entity)
+);
+
+export interface CatalogCategoryVersion<Entity extends CatalogEntity> {
+  name: string;
+  EntityClass: CatalogEntityConstructor<Entity>;
+}
+
+export interface CatalogCategorySpec {
+  group: string;
+  versions: CatalogCategoryVersion<CatalogEntity>[];
+  names: {
+    kind: string;
+  };
+}
+
+export abstract class CatalogCategory extends EventEmitter {
```

Comment:
Maybe split out `CatalogCategory` things to a separate while now that we are moving these anyway.